File: app/src/main/java/com/clockmods/sdk/clock/ClockStyleCapabilities.java
package com.clockmods.sdk.clock;

import java.util.Collections;
import java.util.EnumSet;
import java.util.Set;

/** Declares optional state and motion features understood by a clock style. */
public final class ClockStyleCapabilities {
    public enum Capability {
        SECONDS,
        SMOOTH_SECONDS,
        DATE,
        TIME_ZONE,
        WEATHER,
        STATUS,
        TWENTY_FOUR_HOUR,
        REDUCED_MOTION
    }

    private static final ClockStyleCapabilities NONE =
            new ClockStyleCapabilities(EnumSet.noneOf(Capability.class));

    private final EnumSet<Capability> values;

    private ClockStyleCapabilities(EnumSet<Capability> values) {
        this.values = values.clone();
    }

    public static ClockStyleCapabilities none() {
        return NONE;
    }

    public static ClockStyleCapabilities of(Capability... values) {
        if (values == null || values.length == 0) {
            return none();
        }
        EnumSet<Capability> result = EnumSet.noneOf(Capability.class);
        for (Capability value : values) {
            if (value == null) {
                throw new IllegalArgumentException("Capability must not be null");
            }
            result.add(value);
        }
        return new ClockStyleCapabilities(result);
    }

    public boolean supports(Capability capability) {
        return capability != null && values.contains(capability);
    }

    public Set<Capability> asSet() {
        EnumSet<Capability> copy = EnumSet.noneOf(Capability.class);
        copy.addAll(values);
        return Collections.unmodifiableSet(copy);
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof ClockStyleCapabilities
                && values.equals(((ClockStyleCapabilities) other).values);
    }

    @Override
    public int hashCode() {
        return values.hashCode();
    }

    @Override
    public String toString() {
        return values.toString();
    }
}

File: app/src/main/java/com/clockmods/sdk/clock/ClockStyleCapabilities.kt
package com.clockmods.sdk.clock

import java.util.Collections
import java.util.EnumSet

/** Declares optional state and motion features understood by a clock style. */
class ClockStyleCapabilities private constructor(values: EnumSet<Capability>) {
    enum class Capability {
        SECONDS, SMOOTH_SECONDS, DATE, TIME_ZONE, WEATHER, STATUS, TWENTY_FOUR_HOUR, WORLD_CLOCK,
    }

    private val values = values.clone() as EnumSet<Capability>

    fun supports(capability: Capability?) = capability != null && values.contains(capability)

    fun asSet(): Set<Capability> = Collections.unmodifiableSet(values.clone() as EnumSet<Capability>)

    override fun equals(other: Any?) = other is ClockStyleCapabilities && values == other.values
    override fun hashCode() = values.hashCode()
    override fun toString() = values.toString()

    companion object {
        private val NONE = ClockStyleCapabilities(EnumSet.noneOf(Capability::class.java))

        @JvmStatic
        fun none() = NONE

        @JvmStatic
        fun of(vararg capabilities: Capability?): ClockStyleCapabilities {
            if (capabilities.isEmpty()) return none()
            val result = EnumSet.noneOf(Capability::class.java)
            capabilities.forEach { result.add(requireNotNull(it) { "Capability must not be null" }) }
            return ClockStyleCapabilities(result)
        }
    }
}
